criticality_score/generate.py

```diff
@@ -82,17 +75,34 @@ def get_github_repo_urls_for_language(urls, sample_size, github_lang=None):
                 continue
             urls.append(repo_url)
             new_result = True
-            logger.info(f'Found repository'
-                    f'({samples_processed}): {repo_url}')
+            logger.info(f'{samples_processed} - {repo.name} - {repo_url} - {repo.stargazers_count}')
             samples_processed += 1
             if samples_processed > sample_size:
                 break
         if not new_result:
             break
-        last_stars_processed = repo.stargazers_count
+        upper_limit = repo.stargazers_count
```

Comment:
Correct, I already had this in my main branch. I will add it in the next commit.

But I'd add a smaller number, like 10 for instance.

There are 50K+ repos between 150~250 stars, if no language is specified:
https://github.com/search?q=archived%3Afalse+stars%3A150..250&type=Repositories

For JavaScript, there are 9K+ repos:
https://github.com/search?q=archived%3Afalse+stars%3A150..250+language%3Ajavascript&type=Repositories

What do you think?

Reply:
i feel 10 is too low to cover race, but then we can do those retry , like we originally had to retry everything 3 times, we can do like 2 times ? thoughts?

Reply:
It looks like the approach that should be used for the languages with large number of repos vs smaller ones can be quite different.

CoffeeScript has only 3K+ repos above 10 stars, so adding 50 or 100 stars after each round doesn't affect that much:
https://github.com/search?q=archived%3Afalse+stars%3A%3E10+language%3Acoffeescript&type=Repositories

On the other hand, JavaScript has 25K+ repos only in 10~11 stars range, which makes the situation very different:
https://github.com/search?q=archived%3Afalse+stars%3A10..11+language%3Ajavascript&type=Repositories

I'm not sure what's the best way to cover these cases at the moment. If I will have time, I will try to look at that GraphQL API.

So, for the moment, let me know what works best for you and we can finalize this PR:
- What should be the number to prevent race condition, 50 or 100?
- Fixed lower limit should be 10, or lower?
  - [CoffeeScript - >5 star - 5704 repos](https://github.com/search?q=archived%3Afalse+stars%3A%3E5+language%3Acoffeescript&type=Repositories)
  - [CoffeeScript - >10 star - 3633 repos](https://github.com/search?q=archived%3Afalse+stars%3A%3E10+language%3Acoffeescript&type=Repositories)
- If you wish I can add the "retry" block again?